src/prpy/planning/base.py

```diff
@@ -211,69 +234,50 @@ def get_planners(self, method_name):
+        planners = []
+        results = [None] * len(self._planners)
 
-        planning_threads = list()
-        results = [ None ] * len(self._planners)
-        condition = Condition()
-
-        # Start planning in parallel.
-        for index, planner in enumerate(self._planners):
+        # Find only planners that support the required planning method.
+        for index, planner in enumerate(all_planners):
             if not hasattr(planner, method):
-                results[index] = PlanningError('%s does not implement method %s.' % (planner, method))
+                results[index] = PlanningError(
+                    "{:s} does not implement method {:s}."
+                    .format(planner, method)
+                )
                 continue
+            else:
+                planners.append((index, planner))
 
-            def planning_thread(index, planner):
-                try:
-                    planning_method = getattr(planner, method)
-                    results[index] = planning_method(*args, **kw_args)
-                except MetaPlanningError as e:
-                    results[index] = e
-                except PlanningError as e:
-                    logger.warning('Planning with %s failed: %s', planner, e)
-                    results[index] = e
-
-                # Notify the main thread that we're done.
-                condition.acquire()
-                condition.notifyAll()
-                condition.release()
-
-            thread = Thread(target=planning_thread, args=(index, planner),
-                            name='%s-planning' % planner)
-            thread.daemon = True
-            thread.start()
-
-        # Block until a dominant planner finishes.
-        traj = None
-        condition.acquire()
-        all_done = False
-
-        # TODO: Can we replace this with a fold?
-        while traj is None and not all_done:
-            previous_done = True
-            all_done = True #assume everyone is done
-            for planner, result in zip(self._planners, results):
+        # Helper function to call a planner and store its result or error.
+        def call_planner(index, planner):
+            try:
+                planning_method = getattr(planner, method)
+                kw_args['defer'] = False
+                results[index] = planning_method(*args, **kw_args)
+            except MetaPlanningError as e:
+                results[index] = e
+            except PlanningError as e:
+                logger.warning("Planning with {:s} failed: {:s}"
+                               .format(planner, e))
+                results[index] = e
+
+        # Call every planners in parallel using a concurrent executor and
+        # return the first non-error result in the ordering when available.
+        from trollius.executor import get_default_executor
+        executor = kw_args.get('executor') or get_default_executor()
+        for _ in executor.map(call_planner, planners):
```

Comment:
It waits for the next available result and returns it, until all results are computed.

https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map